src/frameworks/core_graphics/cg_bitmap_context.rs

```diff
@@ -135,9 +167,50 @@ fn put_pixel(
     let (r, g, b, a) = pixel;
     // Blending like this must be done in linear RGB, so this must come before
     // gamma encoding.
-    let (r, g, b) = match data.alpha_info {
-        kCGImageAlphaPremultipliedLast | kCGImageAlphaPremultipliedFirst => (r * a, g * a, b * a),
-        _ => (r, g, b),
+    let (r, g, b, a) = match data.alpha_info {
+        kCGImageAlphaLast => blend_straight(
+            (
+                gamma_decode(pixels[first_component_idx] as f32 / 255.0),
+                gamma_decode(pixels[first_component_idx + 1] as f32 / 255.0),
+                gamma_decode(pixels[first_component_idx + 2] as f32 / 255.0),
+                pixels[first_component_idx + 3] as f32 / 255.0,
+            ),
+            (r, g, b, a),
+        ),
+        kCGImageAlphaFirst => blend_straight(
+            (
+                gamma_decode(pixels[first_component_idx + 3] as f32 / 255.0),
+                gamma_decode(pixels[first_component_idx] as f32 / 255.0),
+                gamma_decode(pixels[first_component_idx + 1] as f32 / 255.0),
+                pixels[first_component_idx + 2] as f32 / 255.0,
```

Comment:
Assuming you want to construct a tuple in (r, g, b, a) order, I think this is wrong? It doesn't match the code below that writes the new pixel out, at least.

Ditto for kCGImageAlphaPremultipliedFirst below.

Reply:
Yeah, the order is incorrect, didn't notice.